oodeel/methods/gram.py

```diff
@@ -0,0 +1,295 @@
+    variable becomes more and more extreme with the sample
+    size. As a result, computing the min/max over the training set is likely to produce
+    min/max values that are so extreme that none of the in distribution correlations of
+    the validation set goes beyond these threshold. The worst is that a significant
+    part of ood data does not exceed the thresholds either. This can be aleviated by
+    computing the min/max over a limited number of sample. However, it is
+    counter-intuitive and, in our opinion, not desirable: adding
+    some more information should only improve a method.
+
+    Hence, we decided to replace the min/max by the q / 1-q quantile, with q a new
+    parameter of the method. Specifically, instead of the deviation as defined in
+    eq. 3 of the paper, we use the definition
+    $$
+    \delta(t_q, t_{1-q}, value) =
+    \begin{cases}
+        0 & \text{if} \; t_q \leq value \leq t_{1-q},  \;\;
+        \frac{t_q - value}{|t_q|} & \text{if } value < t_q,  \;\;
+        \frac{value - t_{1-q}}{|t_q|} & \text{if } value > t_{1-q}
+    \end{cases}
+    $$
+    With this new deviation, the more point we add, the more accurate the quantile
+    becomes. In addition, the method can be made more or less discriminative by
+    toggling the value of q.
+
+    Finally, we found that this approach improved the performance of the baseline in
+    our experiments.
+
+    Args:
+        orders (List[int]): power orders to consider for the correlation matrix
+        quantile (float): quantile to consider for the correlations to build the
+            deviation threshold.
+
+    """
+
+    def __init__(
+        self,
+        orders: List[int] = [i for i in range(1, 11)],
+        quantile: float = 0.01,
+    ):
+        super().__init__()
+        if isinstance(orders, int):
+            orders = [orders]
+        self.orders = orders
+        self.postproc_fns = None
+        self.quantile = quantile
+
+    def _fit_to_dataset(
+        self,
+        fit_dataset: Union[TensorType, DatasetType],
+        val_split: float = 0.2,
+    ) -> None:
+        """
+        Compute the quantiles of channelwise correlations for each layer, power of
+        gram matrices, and class. Then, compute the normalization constants for the
+        deviation. To stay faithful to the spirit of the original method, we still name
+        the quantiles min/max
+
+        Args:
+            fit_dataset (Union[TensorType, DatasetType]): input dataset (ID) to
+                construct the index with.
+            val_split (float): The percentage of fit data to use as validation data for
+                normalization. Default to 0.2.
+        """
+        self.postproc_fns = [
+            self._stat for i in range(len(self.feature_extractor.feature_layers_id))
+        ]
+
+        fit_stats, info = self.feature_extractor.predict(
+            fit_dataset, postproc_fns=self.postproc_fns, return_labels=True
+        )
```

Comment:
If there is only 1 `feature_layer_id`, then `fit_stats` will be a tensor and not a list of tensors at the output of the feature extractor predict function. 

In this case these lines will raise an error:
```python
val_stats = [fit_stat[val_indices] for fit_stat in fit_stats]
fit_stats = [fit_stat[train_indices] for fit_stat in fit_stats]
```

as fit_stat is expected to be a list. The same problem will persist in the `_score_tensor` function where `tensor_stats` is expected to be a list but isn't when there is 1 `feature_layer_id`.

A first solution is to add some conditions on the length of feature_layers_id and to encapsulate the output of the feature extractor when this length is equal to 1 in the code of Gram. An other solution which seems better to me for consistency reasons would be to systematically return a list of tensors in the predict functions of the feature extractors (instead of a tensor when there is 1 output).